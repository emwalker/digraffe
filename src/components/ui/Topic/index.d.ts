import { RelayProp } from 'react-relay';
import { Topic_topic as TopicType } from '__generated__/Topic_topic.graphql';
import { Topic_view as ViewType } from '__generated__/Topic_view.graphql';
declare type Props = {
    orgLogin: string;
    relay: RelayProp;
    topic: TopicType;
    view: ViewType;
};
declare const _default: import("react-relay").Container<Props>;
export default _default;
