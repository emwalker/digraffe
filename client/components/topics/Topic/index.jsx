// @flow
import React from 'react'

type Props = {
  topic: {
    name: string,
    resourcePath: string,
  },
}

export default ({ topic: { name, resourcePath } }: Props) => (
  <li>
    <a href={resourcePath}>{name}</a>
    <div>{resourcePath}</div>
  </li>
)
