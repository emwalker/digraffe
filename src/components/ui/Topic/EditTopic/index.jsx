// @flow
import React, { Component } from 'react'
import { createFragmentContainer, graphql } from 'react-relay'

import Input from '../../Input'
import SaveOrCancel from '../../SaveOrCancel'
import updateTopicMutation from '../../../../mutations/updateTopicMutation'

type Props = {
  id: string,
  isOpen: boolean,
  relay: {
    environment: Object,
  },
  topic: {
    description: string,
    id: string,
    name: string,
  },
  organization: {
    id: string,
  },
  toggleForm: Function,
}

type State = {
  description: string,
  name: string,
}

class EditTopic extends Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      name: props.topic.name,
      description: props.topic.description,
    }
  }

  onSave = () => {
    updateTopicMutation(
      this.props.relay.environment,
      [],
      {
        topicIds: this.addTopicIds,
        description: this.state.description || '',
        id: this.props.topic.id,
        name: this.state.name,
        organizationId: this.props.organization.id,
      },
    )
    this.props.toggleForm()
  }

  // eslint-disable-next-line class-methods-use-this
  get addTopicIds(): string[] {
    return []
  }

  updateDescription = (event: Object) => {
    this.setState({ description: event.currentTarget.value })
  }

  updateName = (event: Object) => {
    this.setState({ name: event.currentTarget.value })
  }

  render() {
    if (!this.props.isOpen)
      return null

    return (
      <div>
        <div className="d-flex col-12">
          <Input
            className="col-5 mr-3"
            id={`edit-link-title-${this.props.id}`}
            label="Name"
            onChange={this.updateName}
            value={this.state.name}
          />
          <Input
            className="col-6"
            id={`edit-topic-description-${this.props.id}`}
            label="Description"
            onChange={this.updateDescription}
            value={this.state.description}
          />
        </div>
        <SaveOrCancel
          onSave={this.onSave}
          onCancel={this.props.toggleForm}
        />
      </div>
    )
  }
}

export default createFragmentContainer(EditTopic, graphql`
  fragment EditTopic_organization on Organization {
    id
  }

  fragment EditTopic_topic on Topic {
    description
    id
    name
  }
`)
