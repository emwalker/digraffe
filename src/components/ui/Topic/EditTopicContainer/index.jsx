// @flow
import React from 'react'
import { QueryRenderer, graphql } from 'react-relay'

import type { TopicType } from 'components/types'
import makeEditTopic from './EditTopic'

type Props = {
  isOpen: boolean,
  orgLogin: string,
  relay: {
    environment: Object,
  },
  toggleForm: Function,
  topic: TopicType,
}

const EditTopicContainer = ({ isOpen, orgLogin, topic, relay, toggleForm }: Props) => (
  <QueryRenderer
    environment={relay.environment}
    query={graphql`
      query EditTopicContainerQuery(
        $viewerId: ID!,
        $orgLogin: String!,
        $repoName: String,
        $repoIds: [ID!],
        $topicId: ID!,
      ) {
        view(
          viewerId: $viewerId,
          currentOrganizationLogin: $orgLogin,
          currentRepositoryName: $repoName,
          repositoryIds: $repoIds,
        ) {
          viewer {
            ...EditTopicForm_viewer
          }

          topic(id: $topicId) {
            ...EditTopic_topic
          }
        }
      }
    `}
    variables={{
      orgLogin,
      repoName: null,
      repoIds: [],
      topicId: topic.id,
      viewerId: '',
    }}
    render={makeEditTopic({ isOpen, orgLogin, relay, toggleForm })}
  />
)

export default EditTopicContainer
