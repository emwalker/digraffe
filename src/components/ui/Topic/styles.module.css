.topicRow :global(.Box-row-link) {
  font-weight: 500;
}
