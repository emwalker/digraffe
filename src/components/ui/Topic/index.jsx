// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import type { Relay, TopicType, UserType, ViewType } from '../../types'
import { liftNodes } from '../../../utils'
import Item from '../Item'
import EditTopic from './EditTopicContainer'

type Props = {
  orgLogin: string,
  relay: Relay,
  topic: TopicType,
  view: ViewType,
  viewer: UserType,
}

type State = {
  formIsOpen: boolean,
}

class Topic extends Component<Props, State> {
  state = {
    formIsOpen: false,
  }

  get repo(): ?Object {
    return this.props.topic.repository
  }

  get currentRepo(): Object {
    return this.props.view.currentRepository
  }

  get topicBelongsToCurrentRepo(): boolean {
    if (!this.repo)
      return true
    return this.repo.id === this.currentRepo.id
  }

  get displayColor(): string {
    return this.topicBelongsToCurrentRepo
      ? 'transparent'
      : this.props.topic.repository.displayColor
  }

  get parentTopics(): TopicType[] {
    return liftNodes(this.props.topic.parentTopics)
  }

  get showEditButton(): boolean {
    return !this.props.topic.loading && !this.props.viewer.isGuest
  }

  toggleForm = () => {
    this.setState(({ formIsOpen }) => ({ formIsOpen: !formIsOpen }))
  }

  render = () => {
    const { topic } = this.props

    return (
      <Item
        className="Box-row--topic"
        description={this.props.topic.description}
        displayColor={this.displayColor}
        formIsOpen={this.state.formIsOpen}
        newlyAdded={this.props.topic.newlyAdded}
        orgLogin={this.props.orgLogin}
        repoName={topic.repository && topic.repository.name}
        showEditButton={this.showEditButton}
        showLink={false}
        title={topic.name}
        toggleForm={this.toggleForm}
        topics={this.parentTopics}
        url={topic.resourcePath}
      >
        <EditTopic
          isOpen={this.state.formIsOpen}
          orgLogin={this.props.orgLogin}
          relay={this.props.relay}
          toggleForm={this.toggleForm}
          topic={topic}
          view={this.props.view}
        />
      </Item>
    )
  }
}

export default createFragmentContainer(Topic, graphql`
  fragment Topic_view on View {
    currentRepository {
      id
    }
  }

  fragment Topic_viewer on User {
    isGuest
  }

  fragment Topic_topic on Topic {
    description
    id
    loading
    name
    newlyAdded
    resourcePath

    repository {
      name
      displayColor
      id
    }

    parentTopics(first: 10) {
      edges {
        node {
          name
          resourcePath
        }
      }
    }
  }
`)
