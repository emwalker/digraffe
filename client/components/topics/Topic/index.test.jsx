import React from 'react'
import { shallow } from 'enzyme'
import Topic from './index'

describe('<Topic />', () => {
  const topic = {
    name: 'Frotz',
    resourcePath: '/topics/1234',
    id: '1234',
  }

  const wrapper = shallow(<Topic topic={topic} />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
