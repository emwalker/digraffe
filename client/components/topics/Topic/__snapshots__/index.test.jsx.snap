// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Topic /> renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    topic={
      Object {
        "id": "1234",
        "name": "Frotz",
        "resourcePath": "/topics/1234",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <a
          href="/topics/1234"
        >
          Frotz
        </a>,
        <div>
          /topics/1234
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Frotz",
          "href": "/topics/1234",
        },
        "ref": null,
        "rendered": "Frotz",
        "type": "a",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "/topics/1234",
        },
        "ref": null,
        "rendered": "/topics/1234",
        "type": "div",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <a
            href="/topics/1234"
          >
            Frotz
          </a>,
          <div>
            /topics/1234
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Frotz",
            "href": "/topics/1234",
          },
          "ref": null,
          "rendered": "Frotz",
          "type": "a",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "/topics/1234",
          },
          "ref": null,
          "rendered": "/topics/1234",
          "type": "div",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
